product/ci/addon/tasks/runtests.go

```diff
@@ -339,29 +342,44 @@ func (r *runTestsTask) getTestSelection(ctx context.Context, files []types.File,
 
 	if isManual {
 		// Manual execution: Select all tests in case of manual execution
-		log.Infow("Detected manual execution - for test intelligence to be configured, a PR must be raised. Running all the tests")
-		r.runOnlySelectedTests = false
-	} else if len(files) == 0 {
-		// PR execution: Select all tests if unable to find changed files list
-		log.Infow("Unable to get changed files list")
+		log.Infow("Detected manual execution - for test intelligence to be configured the execution should be via a PR or Push trigger, running all the tests.")
 		r.runOnlySelectedTests = false
 	} else {
-		// PR execution: Call TI svc only when there is a chance of running selected tests
-		var err error
-		resp, err = selectTestsFn(ctx, files, r.runOnlySelectedTests, r.id, r.log, r.fs)
-		if err != nil {
-			log.Errorw("There was some issue in trying to intelligently figure out tests to run. Running all the tests", "error", zap.Error(err))
-			r.runOnlySelectedTests = false
-		} else if !valid(resp.Tests) { // This shouldn't happen
-			log.Errorw("Test Intelligence did not return suitable tests")
-			r.runOnlySelectedTests = false
-		} else if resp.SelectAll == true {
-			log.Infow("Test Intelligence determined to run all the tests")
+		var errChangedFiles error = nil
+		if isPushTriggerFn() {
+			lastSuccessfulCommitID, err := getCommitInfoFn(ctx, r.id, r.log)
+			if err != nil || lastSuccessfulCommitID == "" {
+				log.Infow("Test Intelligence determined to run all tests to bootstrap", "error", zap.Error(err))
+				r.runOnlySelectedTests = false // TI selected all the tests to be run
+				return resp
+			}
+			log.Infof("Using reference commit: %s\n", lastSuccessfulCommitID)
+			files, errChangedFiles = getChangedFilesPushTriggerFn(ctx, r.id, lastSuccessfulCommitID, r.log)
+		}
+		if errChangedFiles != nil || len(files) == 0 {
+			// PR execution: Select all tests if unable to find changed files list
+			log.Infow("Unable to get changed files list")
 			r.runOnlySelectedTests = false
 		} else {
-			r.log.Infow(fmt.Sprintf("Running tests selected by Test Intelligence: %s", resp.Tests))
+			files = utils.ReadPkgs(r.log, r.fs, files)
```

Comment:
Implement an interface similar to VM lite-engine. This won't be scalable for other languages / tools
```suggestion
			filesWithPkg := runner.ReadPackages(workspace, files)
```

Reply:
Done